Custom error documentation (#10937)
ethereum/solidity: docs/abi-spec.rst

efficient search and arbitrary legibility by defining events with two arguments — one
indexed, one not — intended to hold the same value.

.. _abi_errors:

Errors
======

In case of a failure inside a contract, the contract can use a special opcode to abort execution and revert
all state changes. In addition to these effects, descriptive data can be returned to the caller.
This descriptive data is the encoding of an error plus arguments in the same way as data for a function
call would be encoded.

As an example, let us consider the following contract whose ``transfer`` function always
reverts with a custom error of "insufficient balance":

::

    // SPDX-License-Identifier: GPL-3.0

>     pragma solidity ^0.8.1;

Reviewer: ```suggestion
    pragma solidity ^0.8.2;
```

0.8.1 is also used in a few other places.

    contract TestToken {
        error InsufficientBalance(uint256 available, uint256 required);

>         function transfer(address to, unit amount) {

Reviewer 2: ```suggestion
        function transfer(address to, uint amount) {
```

>           revert(InsufficientBalance(0, amount));

Reviewer: ```suggestion
            revert(InsufficientBalance(0, amount));
```

        }
    }

The return data would be encoded in the same way as a call to a function

> ``InsufficientBalance(0, amount)``, i.e. ``0xcf479181``, ``u256(0)``, ``amount``.

Reviewer 2: ```suggestion
``InsufficientBalance(0, amount)``, i.e. ``0xcf479181``, ``u256(0)``, ``u256(amount)``.
```

> ``InsufficientBalance(0, amount)``, i.e. ``0xcf479181``, ``u256(0)``, ``amount``.

Reviewer: The only place in the docs where I have seen any mention of type called `u256` is the page about Yul. I think this is written from the perspective of Solidity so it should use `uint256` instead?

```suggestion
``InsufficientBalance(0, amount)``, i.e. ``0xcf479181``, ``uint256(0)``, ``amount``.
```

efficient search and arbitrary legibility by defining events with two arguments — one
indexed, one not — intended to hold the same value.

.. _abi_errors:

Errors
======

In case of a failure inside a contract, the contract can use a special opcode to abort execution and revert
all state changes. In addition to these effects, descriptive data can be returned to the caller.

> This descriptive data is the encoding of an error plus arguments in the same way as data for a function

Reviewer 3: ```suggestion
This descriptive data is the encoding of an error and its arguments in the same way as data for a function
```

call would be encoded.

As an example, let us consider the following contract whose ``transfer`` function always
reverts with a custom error of "insufficient balance":

::

    // SPDX-License-Identifier: GPL-3.0
    pragma solidity ^0.8.3;

    contract TestToken {
        error InsufficientBalance(uint256 available, uint256 required);
        function transfer(address /*to*/, uint amount) public pure {
            revert InsufficientBalance(0, amount);
        }
    }

The return data would be encoded in the same way as the function call

> ``InsufficientBalance(0, amount)``, i.e. ``0xcf479181``, ``uint256(0)``, ``amount``.

Reviewer 4: Maybe better to write also what is exactly used to calculate the selector
```
same way as the function call ``InsufficientBalance(0, amount)``, i.e. selector of 
``InsufficientBalance(uint256,uint256)`` -> ``0xcf479181``, ...
```

> The error selectors ``0x00000000`` and ``0xffffffff`` are reserved for future use.

Reviewer 3: Aren't they already used?

Author: No, function selectors might but not error selectors.

efficient search and arbitrary legibility by defining events with two arguments — one
indexed, one not — intended to hold the same value.

.. _abi_errors:

Errors
======

In case of a failure inside a contract, the contract can use a special opcode to abort execution and revert
all state changes. In addition to these effects, descriptive data can be returned to the caller.

> This descriptive data is the encoding of an error and its arguments in the same way as data for a function

Reviewer 4: `.. and its arguments are encoded in the same way as data for a function call.`

call would be encoded.

As an example, let us consider the following contract whose ``transfer`` function always
reverts with a custom error of "insufficient balance":

::

    // SPDX-License-Identifier: GPL-3.0
    pragma solidity ^0.8.3;

    contract TestToken {
        error InsufficientBalance(uint256 available, uint256 required);
        function transfer(address /*to*/, uint amount) public pure {
            revert InsufficientBalance(0, amount);
        }
    }

The return data would be encoded in the same way as the function call
``InsufficientBalance(0, amount)`` to the function ``InsufficientBalance(uint256,uint256)``,
i.e. ``0xcf479181``, ``uint256(0)``, ``uint256(amount)``.

The error selectors ``0x00000000`` and ``0xffffffff`` are reserved for future use.

Note that since error data in inner calls is just "handed up", it can happen that
a smart contract causes errors that are not defined in the contract itself or
any of its files. Furthermore, never trust error data because it can be "forged"

> in an inner call.

Reviewer: I'd make it a warning and also reword to explain what it means to forge an error:

```suggestion
.. warning::

    Never trust error data.
    The error data by default bubbles up through the chain of external calls, which
    means that a contract may receive an error not defined in any of the contracts
    it calls directly.
    Furthermore, any contract can fake any error by returning data that matches
    an error signature, even if the error is not defined anywhere.
```

- ``anonymous``: ``true`` if the event was declared as ``anonymous``.

Errors look as follows:

- ``type``: always ``"error"``
- ``name``: the name of the error.
- ``inputs``: an array of objects, each of which contains:

  * ``name``: the name of the parameter.
  * ``type``: the canonical type of the parameter (more below).
  * ``components``: used for tuple types (more below).

.. note::

>   There can be multiple errors with the same name and even with identical signature

Reviewer 4: Maybe we should add a note here with something like `note: errors - even if defined in different contracts/libraries - are placed in the same (global) namespace`. We should also add the same note in the event section. I think that would it make more clear.

Author: I added some more text.
